invenio_rdm_records/services/components.py

```diff
@@ -144,13 +148,75 @@ def new_version(self, identity, draft=None, record=None, **kwargs):
 class ExternalPIDsComponent(ServiceComponent):
     """Service component for pids."""
 
+    def get_provider(self, scheme):
+        """Process a given PID with its provider."""
+        provider = self.service.config.pids_provides.get(scheme)
+
+        if not provider:
+            raise ValidationError(
+                _(f"No provider configured for PID type {scheme}"),
+                field_name="pids"
+            )
+
+        return provider
+
     def create(self, identity, data=None, record=None, **kwargs):
         """Inject parsed pids to the record."""
-        record.pids = data.get('pids', {})
+        pids = data.get('pids', {})
+
+        for scheme, pid_attrs in pids.items():
+            if pid_attrs.get("provider") != PID_PROVIDER_EXTERNAL:
+                provider = self.get_provider(scheme)
+
+                # PIDS-FIXME: Do we assume it fails, or should we re-raise
+                # with ValidationErrors like AccessCompoenent (I think yes)
+                provider.validate(**pid_attrs)
+
+                # NOTE: When reserving a PID the record/draft was created
+                # and therefore any following action is an update
+                if pid_attrs.get("identifier"):
+                    raise ValidationError(
+                        _("Cannot create managed PID with existing value."),
+                        field_name="pids"
+                    )
+
+                pid = provider.create(status=PIDStatus.RESERVED)
+                pid_attrs["identifier"] = pid.value
+                pids[scheme] = pid_attrs
+
+            elif not pid_attrs.get("identifier"):
+                raise ValidationError(
+                    _("Unmanaged PID must contain an identifier value"),
+                    field_name="pids"
+                )
+
+        record.pids = pids
 
     def update_draft(self, identity, data=None, record=None, **kwargs):
         """Inject parsed pids to the record."""
-        record.pids = data.get('pids', {})
+        pids = data.get('pids', {})
+
+        for scheme, pid_attrs in pids.items():
+            if pid_attrs.get("provider") != PID_PROVIDER_EXTERNAL:
+                provider = self.get_provider(scheme)
+
+                # NOTE: value format is checked by the marshmallow schema
+                provider.validate(**pid_attrs)
+
+                identifier_value = pid_attrs.get("identifier")
```

Comment:
I am not sure we should do this when saving... here my scenario:
1. click on new upload btn
2. input a title
3. hit the Save Draft button
4. suddenly the DOI has been reserved and appears in the deposit form

I think we might want to do that when publishing only?

Reply:
I think that is the difference between "reserve" and "register" and its responsability is even delegated to the Provider (so implementation-dependent). The way I see it:

1 and 2 stay the same
3. Upon save, the provider will be implemented in a way that the reserve only happens locally (i.e. just save in PIDStore).
4. A DOI will appear in the deposit form, but is not available in DataCite

I am not sure if your concern is the fact of reserve goint to DataCite, or also not creating a `DOI` in PIDStore. On the other hand, if we do as you say, there will be less "unused" pids due to random saves of drafts that went nowhere afterwards...

Reply:
I think I was not able to explain myself :) Let's have a chat about this :)